### PR TITLE
[SPARK-21403][Mesos] fix --packages for mesos

### PR DESCRIPTION
**Pull request by skonto**
## What changes were proposed in this pull request?
Fixes --packages flag for mesos in cluster mode. Probably I will handle standalone and Yarn in another commit, I need to investigate those cases as they are different.

## How was this patch tested?
Tested with a community 1.9 dc/os cluster. packages were successfully resolved in cluster mode within a container.

@andrewor14  @susanxhuynh @ArtRand @srowen  pls review.

### CODE REVIEW

File: core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala [hunk 1]
```diff
@@ -269,6 +269,25 @@ object SparkSubmit extends CommandLineUtils {
...
+        printErrorAndExit("Cluster deploy mode is not compatible with master \"local\"")
+      case (_, CLUSTER) if isShell(args.primaryResource) =>
+        printErrorAndExit("Cluster deploy mode is not applicable to Spark shells.")
+      case (_, CLUSTER) if isSqlShell(args.mainClass) =>
+        printErrorAndExit("Cluster deploy mode is not applicable to Spark SQL shell.")
+      case (_, CLUSTER) if isThriftServer(args.mainClass) =>
+        printErrorAndExit("Cluster deploy mode is not applicable to Spark Thrift server.")
+      case _ =>
+    }
+
```
Comment (skonto) [inline core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala]: This is for failing faster no need to resolve dependencies if we are going to fail.

**Comment by vanzin** [comment]
I'm 99% sure there's nothing to do for YARN. This line takes care of it:

    args.jars = mergeFileLists(args.jars, resolvedMavenCoordinates)

YARN cluster mode will distribute all jars in `args.jars` to the app.

As for the change, it seems to work because the Mesos backend starts the driver using `spark-submit`, right? (It would probably have to change the deploy mode from "cluster" to "client" when doing that but I didn't dig that much into the code...)

If that's the case it seems fine, although it kinda loses the ability to use the ivy cache on the machine launching the job...

Also, I'd be more comfortable if someone more familiar with the Mesos backend could take a look. Not sure who that person is.

**Comment by skonto** [comment]
@vanzin 
> args.jars = mergeFileLists(args.jars, resolvedMavenCoordinates)

My understanding:
Files intended for the appmaster are handled by copying to hdfs:
https://github.com/apache/spark/blob/1cad31f00644d899d8e74d58c6eb4e9f72065473/resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala#L365
The distributed cache manager does the work by utilizing the distributed Hadoop cache for jars at the executor side. 
https://github.com/apache/spark/blob/ab9872db1f9c0f289541ec5756d1a142d85545ce/resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/ClientDistributedCacheManager.scala

> As for the change, it seems to work because the Mesos backend starts the driver using spark-submit, right?
That is the idea. The second time the submit is called in order to launch the driver, the submit is done using client mode which is the default. The call for launching the driver in client mode is here:
https://github.com/apache/spark/blob/8da3f7041aafa71d7596b531625edb899970fec2/resource-managers/mesos/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosClusterScheduler.scala#L411-L442

> If that's the case it seems fine, although it kinda loses the ability to use the ivy cache on the machine launching the job...

That's is correct. If you launch jobs from a machine within the same network cache makes sense. Otherwise it is just copying over the internet. Now in restarts it is also important when you re-launch something to access it from the cache. In spark there is no functionality for the mesos code to exploit any type of cache from what I see. I would prefer a unified cluster layer for certain things like in the case of other frameworks: https://issues.apache.org/jira/browse/FLINK-6177
I guess some refactoring would make yarn stuff accessible from the mesos part in Spark. Some other options are the fetcher cache mechanism but not so sure if it is distributed, it is only local per agent, need to discuss it further with @susanxhuynh @ArtRand

**Comment by vanzin** [comment]
Sounds good. Merging to master.

**Comment by skonto** [comment]
thnx @vanzin I am working on some code for the standalone case.

**Pull request closed without merging**